File: python/3/part2.py
class Solve:
    def __init__(self):
        with open("input.txt", "r") as F:
            self.file = [line.strip() for line in F.readlines()]

        self.suma = 0

        self.run()
        print(self.suma)

    def get_number_from(self, i: int, j: int) -> int:
        """Pozyskuje liczbę, która zawiera w sobie punkt o współrzędnych i oraz j"""
        number = self.file[i][j]

        # Warunek, mówiący w jakim kierunku odkrywamy liczbę
        for go_left in range(0, 2):
            shift = -1 if go_left else 1
            
            # Zabezpieczenie przed wyjściem poza zakres
            if j + shift < 0 or j + shift >= len(self.file[i]):
                continue

            while self.file[i][j+shift].isdigit():
                if go_left:
                    number = self.file[i][j+shift] + number

                else:
                    number += self.file[i][j+shift]

                shift += -1 if go_left else 1

                # Zabezpieczenie przed wyjściem poza zakres
                if j + shift < 0 or j + shift >= len(self.file[i]):
                    break
        
        return int(number)

    def find_numbers_around(self, i: int, j: int) -> None:
        """Sprawdź czy wokół punktu znajdują się liczby, jeżeli tak - pozyskaj je"""
        digits_found = set()

        # Dla linijki ponad, na tym samym poziomie i poniżej
        for height in range(-1, 2):
            # Jeżeli jesteśmy w pierwszym wierszu, albo w ostatnim to skip
            if i <= 0 or i+height >= len(self.file):
                continue


            # Sprawdź czy w obecnie oprzeszukiwanej linijce znajdują się liczby
            for shift in range(j if j == 0 else j-1, j+2): # j+2 ponieważ range() to zbiór lewostronnie domknięty
                for letter in self.file[i+height][shift]:
                    if letter.isdigit():
                        digits_found.add(self.get_number_from(i+height, shift))

        # Jeżeli znaleziono dwie liczby wokół `*` to je przemnóż i dodaj do sumy
        if len(digits_found) == 2:
            out = 1

            for n in digits_found:
                out *= n

            self.suma += out
                
    def run(self):
        """Główna funkcja biegająca"""
        # Dla rzędu
        for i in range(0, len(self.file)):
            # Dla kolumny
            for j in range(0, len(self.file[i])):
                if self.file[i][j] == "*":
                    self.find_numbers_around(i, j)


if __name__ == "__main__":
    Solve()

File: c++/3/part2.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <set>
#include <cctype>

using std::vector;
using std::string;
using std::cout;
using std::endl;
using std::isdigit;

class Solve {
private:
	int sum = 0;
	vector<string> file;
	
	void main() {
		for (int i = 0; i < file.size(); i++) {
			for (int j = 0; j < file[i].length(); j++) {
				if (file[i][j] == '*') {
					find_numbers_around(i, j);
				}
			}
		}
	}

	void find_numbers_around(int i, int j) {
		std::set<int> digits_found;

		for (int height = -1; height < 2; height++) {
			if (i <= 0 || i + height >= file.size()) {
				continue;
			}

			int shift = j-1;

			if (j == 0) {
				shift++;
			}

			for (; shift < j+2; shift++) {
				if (isdigit(file[i+height][shift])) {
					digits_found.insert(get_number_from(i+height, shift));
				}
			}

		}

		if (digits_found.size() == 2) {
			int m = 1;
			for (int digit : digits_found) {
				m *= digit;
			}
			sum += m;	
		}
	}

	int get_number_from(int i, int j) {
		string number = "";
		number.push_back(file[i][j]);

		for (int go_left = 0; go_left < 2; go_left++) {
			int shift = 1;

			if (go_left) {
				shift = -1;
			}

			if (j + shift < 0 || j + shift >= file[i].length()) {
				continue;
			}

			while (isdigit(file[i][j+shift])) {
				if (go_left) {
					number = file[i][j+shift] + number;
				} else {
					number.push_back(file[i][j+shift]);
				}

				if (go_left) {
					shift--;
				} else {
					shift++;
				}

				if (j + shift < 0 || j + shift >= file[i].length()) {
					break;
				}
			}

		}

		return std::stoi(number);
	}

public:
	Solve() {
		std::ifstream input("input.txt");
		string line;

		while (std::getline(input, line)) {
			file.push_back(line);
		}

		input.close();
		main();
		cout << sum << endl;	
	}	
};

int main() {
	Solve();

	return 0;
}
